core/include/detray/propagator/actors/pointwise_material_interactor.hpp

```diff
@@ -125,73 +127,95 @@ struct pointwise_material_interactor : actor {
+    /// @param[in]  e_loss energy loss
+    /// @param[in]  sign navigation direction
+    DETRAY_HOST_DEVICE inline void update_qop(
+        bound_vector &vector, const scalar_type p, const scalar_type q,
+        const scalar_type m, const scalar_type e_loss, const int sign) const {
+
+        // Get new Energy
+        const scalar_type nextE{
+            std::sqrt(m * m + p * p) -
+            std::copysign(e_loss, static_cast<scalar_type>(sign))};
+
+        // Put particle at rest if energy loss is too large
+        const scalar_type nextP{(m < nextE) ? std::sqrt(nextE * nextE - m * m)
+                                            : 0.f};
+
+        // For neutral particles, qoverp = 1/p
+        getter::element(vector, e_bound_qoverp, 0) =
+            (q != 0.f) ? q / nextP : 1.f / nextP;
+    }
 
-            // variance of scattering angle
-            const scalar_type var_scattering_angle{
-                std::copysign(s.scattering_angle * s.scattering_angle,
-                              static_cast<scalar_type>(sign))};
+    /// @brief Update the variance of q over p of bound track parameter
+    ///
+    /// @param[out] covariance covariance matrix of bound track parameter
+    /// @param[in]  sigma_qop variance of q over p
+    /// @param[in]  sign navigation direction
+    DETRAY_HOST_DEVICE inline void update_qop_variance(
+        bound_matrix &covariance, const scalar_type sigma_qop,
+        const int sign) const {
 
-            const auto dir = stepping._bound_params.dir();
+        const scalar_type variance_qop{sigma_qop * sigma_qop};
 
-            matrix_operator().element(covariance, e_bound_phi, e_bound_phi) +=
-                var_scattering_angle / (dir[0] * dir[0] + dir[1] * dir[1]);
+        matrix_operator().element(covariance, e_bound_qoverp, e_bound_qoverp) +=
+            std::copysign(variance_qop, static_cast<scalar_type>(sign));
+    }
 
-            matrix_operator().element(covariance, e_bound_theta,
-                                      e_bound_theta) += var_scattering_angle;
-        }
+    /// @brief Update the variance of phi and theta of bound track parameter
+    ///
+    /// @param[out] covariance covariance matrix of bound track parameter
+    /// @param[in]  dir direction of track
+    /// @param[in]  projected_scattering_angle projected scattering angle
+    /// @param[in]  sign navigation direction
+    DETRAY_HOST_DEVICE inline void update_angle_variance(
+        bound_matrix &covariance, const vector3 &dir,
+        const scalar_type projected_scattering_angle, const int sign) const {
+
+        // variance of projected scattering angle
+        const scalar_type var_scattering_angle{std::copysign(
+            projected_scattering_angle * projected_scattering_angle,
+            static_cast<scalar_type>(sign))};
+
+        matrix_operator().element(covariance, e_bound_phi, e_bound_phi) +=
+            var_scattering_angle / (1 - dir[2] * dir[2]);
```

Comment:
I just saw this now, but is this different to what was calculated before?
```
matrix_operator().element(covariance, e_bound_phi, e_bound_phi) +=
    var_scattering_angle / (dir[0] * dir[0] + dir[1] * dir[1]);
```

Reply:
It is the same because dir is unit vector. So the new one improves the speed performance

Reply:
Ah, Ok, thanks